src/backend/distributed/transaction/transaction_recovery.c

```diff
@@ -346,6 +350,7 @@ RecoverWorkerTransactions(WorkerNode *workerNode)
 
 	systable_endscan(scanDescriptor);
 	table_close(pgDistTransaction, NoLock);
+	UnlockShardResource((uint64) distTransactionRelationId, ShareUpdateExclusiveLock);
```

Comment:
I don't think you should unlock here. That allows another recovery to proceed before changes are committed.